File: 1/2020.10.17/78.Subsets/java/solution.java
class Solution {

    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtracking(nums, 0, new ArrayList<>());
        return res;
    }

    private void backtracking(int[] nums, int start, List<Integer> tmp) {
        res.add(new ArrayList<>(tmp));
        for (int i = start; i < nums.length; i++) {
            tmp.add(nums[i]);
            backtracking(nums, i + 1, tmp);
            tmp.remove(tmp.size() - 1);
        }
    }
}

File: 1/2020.10.17/78.Subsets/python/solution.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def backtracking(start, tmp):
            res.append(tmp[:])
            for i in range(start, len(nums)):
                tmp.append(nums[i])
                backtracking(i + 1, tmp)
                tmp.pop()
        res = []
        backtracking(0, [])
        return res

    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for num in nums:
            res += [i + [num] for i in res]
        return res
